xarray/backends/common.py

```diff
@@ -344,12 +344,13 @@ def encode(self, variables, attributes):
 
 
 class BackendEntrypoint:
-    __slots__ = ("guess_can_open", "open_dataset", "open_dataset_parameters")
+    open_dataset_parameters: Union[Tuple, None] = None
 
-    def __init__(self, open_dataset, open_dataset_parameters=None, guess_can_open=None):
-        self.open_dataset = open_dataset
-        self.open_dataset_parameters = open_dataset_parameters
-        self.guess_can_open = guess_can_open
+    def open_dataset(self):
+        raise NotImplementedError
 
+    def guess_can_open(self, store_spec):
+        return False
 
-BACKEND_ENTRYPOINTS: Dict[str, BackendEntrypoint] = {}
+
+BACKEND_ENTRYPOINTS: Dict[str, Type[BackendEntrypoint]] = {}
```

Comment:
from a typing perspective, wont this will require all backends to subclass from `BackendEntrypoint`? So while we seem to have decided to not to explicitly check for inheritance, this would seem to require it.

Reply:
Mm.. you are right, and in fact this is a point in favour of inheritance. So we can control typing. Otherwise that would be `Any` that feels less correct to me.

Reply:
`typing.Protocol` seems to be the best solution, but I would avoid adding new requirements only for this.